src/shared/xen-util.c

```diff
@@ -0,0 +1,45 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#if HAVE_XENCTRL
+#  include <xenctrl.h>
+#endif
+
+#include "dlfcn-util.h"
+#include "strv.h"
+#include "xen-util.h"
+
+#if HAVE_XENCTRL
+static void* xenctrl_dl = NULL;
+
+xc_interface * (*sym_xc_interface_open)(xentoollog_logger *, xentoollog_logger *, unsigned) = NULL;
+int (*sym_xc_kexec_status)(xc_interface *, int) = NULL;
+int (*sym_xc_interface_close)(xc_interface *) = NULL;
+#endif
+
+int dlopen_xenctrl(void) {
+#if HAVE_XENCTRL
+        int r;
+
+        /* There was no ABI change for those functions over these releases */
+        FOREACH_STRING(s, "libxenctrl.so.4.16",
+                          "libxenctrl.so.4.15",
+                          "libxenctrl.so.4.14",
+                          "libxenctrl.so.4.13",
+                          "libxenctrl.so.4.12",
+                          "libxenctrl.so.4.11",
+                          "libxenctrl.so.4.10",
+                          "libxenctrl.so.4.9") {
```

Comment:
yes, this library is happy to break ABIs!

Reply:
those are not ABI breaks, looks like a build/packaging issue - instead of shipping the full library as `libxenctrl.so.4.14.0` and the symlink as `libxenctrl.so.4`, the symlink includes the ABI revision as well, for some reason, which gets bumped on every new symbol

nothing we can do about it here, except maybe look for `libxenctrl.so.4` first in case it ever gets added?

Reply:
@sthibaul do you think there's any chance this will ever get solved? I'd add it regardless, it's cheap enough in this context to look for one extra file

Reply:
It's not "bumped on every new symbol", it's simply the version of Xen.

I don't think they will ever add a `.4` symlink: they do break ABI every now and then. Not for simple calls like `xc_kexec_status`, but for more involved calls like creating a domain, yes.

Reply:
Any chance you could ask? As you can see here, it's going to be a massive pain to keep this list up to date, and none of us use Xen anywhere so I'd be up to you to send PRs to update it every time they release a new version...

Reply:
Should we be smarter about finding the SO? There's nothing preventing us from figuring out the installed SO in meson and setting a define and looking that up at runtime.

Reply:
One still has to check on each release that the ABI hasn't been changed. Even if that's little probable, it may happen.

Reply:
> Any chance you could ask?

Really, there is no chance that they add such a link. They really don't promise any ABI stability for xenctrl.